esrally/telemetry.py

```diff
@@ -568,6 +568,134 @@ def record(self):
+        else:
+            self.indices_per_cluster = indices_per_cluster
+
+        if self.indices_per_cluster:
+            for cluster_name in self.indices_per_cluster.keys():
+                if cluster_name not in clients:
+                    raise exceptions.SystemSetupError(
+                        "The telemetry parameter 'shard-stats-indices' must be a JSON Object with keys matching "
+                        "the cluster names [{}] specified in --target-hosts "
+                        "but it had [{}].".format(",".join(sorted(clients.keys())), cluster_name))
+            self.specified_cluster_names = self.indices_per_cluster.keys()
+
+        self.metrics_store = metrics_store
+        self.samplers = []
+
+    def on_benchmark_start(self):
+        for cluster_name in self.specified_cluster_names:
+            recorder = ShardStatsRecorder(cluster_name, self.clients[cluster_name], self.metrics_store,
+                                             self.sample_interval,
+                                             self.indices_per_cluster[cluster_name] if self.indices_per_cluster else "")
+            sampler = SamplerThread(recorder)
+            self.samplers.append(sampler)
+            sampler.setDaemon(True)
+            # we don't require starting recorders precisely at the same time
+            sampler.start()
+
+    def on_benchmark_stop(self):
+        if self.samplers:
+            for sampler in self.samplers:
+                sampler.finish()
+
+
+class ShardStatsRecorder:
+    """
+    Collects and pushes shard stats for the specified cluster to the metric store.
+    """
+
+    def __init__(self, cluster_name, client, metrics_store, sample_interval, indices=None):
+        """
+        :param cluster_name: The cluster_name that the client connects to, as specified in target.hosts.
+        :param client: The Elasticsearch client for this cluster.
+        :param metrics_store: The configured metrics store we write to.
+        :param sample_interval: integer controlling the interval, in seconds, between collecting samples.
+        :param indices: optional list of indices to filter results from.
+        """
+
+        self.cluster_name = cluster_name
+        self.client = client
+        self.metrics_store = metrics_store
+        self.sample_interval = sample_interval
+        self.indices = indices
+        self.logger = logging.getLogger(__name__)
+
+    def __str__(self):
+        return "shard stats"
+
+    def record(self):
+        """
+        Collect _cat/shards for indexes (optionally) specified in telemetry parameters and push to metrics store.
+        """
+        # pylint: disable=import-outside-toplevel
+        import elasticsearch
+        try:
+            stats = self.client.cat.shards(index=self.indices)
+        except elasticsearch.TransportError:
+            msg = "A transport error occurred while collecting _cat/shards on cluster [{}]".format(self.cluster_name)
```

Comment:
We should use f-strings for all new code.

Reply:
Have you considered using https://github.com/asottile/pyupgrade as part of `make lint` to enforce f-strings everywhere? There are ways to [avoid ruining git blame](https://github.com/psf/black/#migrating-your-code-style-without-ruining-git-blame) if that's a concern.

Reply:
Ah, thanks for the idea; I wasn't aware of that. [pyupgrade](https://github.com/asottile/pyupgrade) could really be a nice approach to migrate to f-strings across the code base. We already discussed a while back using black for automatic formatting in another tool that we maintain but did not get around to introducing it yet so mid-term this could indeed be an option.